flips/20220203-capability-controllers.md

```diff
@@ -0,0 +1,544 @@
+so that each storage path on an account has an array of CapCons of Capabilities issued from that storage path. 
+CapCons are a non-storable object (similar to AuthAccounts).
+
+The definition of the `CapabilityController`.
+
+```cadence
+struct CapabilityController {
+    /// The block height when the controlled capability was created.
+    let issueHeight: UInt64
+   
+    /// The type of the controlled capability, i.e. the T in `Capability<T>`.
+    let borrowType: Type
+   
+    /// The identifier of the controlled capability.
+    /// All copies of a capability have the same ID.
+    let capabilityID: UInt64
+   
+    /// Returns the targeted storage path of the controlled capability.
+    fun target(): StoragePath
+
+    /// Delete this capability controller,
+    /// and disable the controlled capability and its copies.
+    ///
+    /// The controller will be deleted from storage,
+    /// but the controlled capability and its copies remain.
+    ///
+    /// Once this function returns, the controller is no longer usable,
+    /// all further operations on the controller will panic.
+    ///
+    /// Borrowing from the controlled capability or its copies will return nil.
+    ///
+    fun delete()
+   
+    /// Retarget the capability.
+    /// This moves the CapCon from one CapCon array to another.
+    fun retarget(target: StoragePath)
+
+}
+```
+
+Capabilities would expose the new field `let id: UInt64`, in addition to the account field they already expose.
+
+The capability related methods in the `AuthAccount` and in the `PublicAccount` would be moved to a `capabilities` namespace, 
+similar to how the contract methods are in the [contracts namespace](https://developers.flow.com/cadence/language/accounts).
+
+The `AuthAccount` would get new methods to create or get or iterate through capabilities,
+and to get or iterate through CapCons in order to manage capabilities. 
+Methods used for (un)linking would be removed as they are no longer needed.
+
+```cadence
+struct AuthAccount {
+    // ...
+    // removed: fun link<T: &Any>(_ newCapabilityPath: CapabilityPath, target: Path): Capability<T>?
+    // removed: fun getLinkTarget(_ path: CapabilityPath): Path?
+    // removed: fun unlink(_ path: CapabilityPath)
+    // moved & renamed: fun getCapability<T>(_ path: CapabilityPath): Capability<T>
+    // moved & renamed: fun forEachPublic(_ function: ((PublicPath, Type): Bool))
+
+    let capabilities: AuthAccount.Capabilities
+
+    struct Capabilities {
+        /// get returns the capability at the public path, if one was stored there.
+        fun get<T>(_ path: PublicPath): Capability<T>?
+
+        /// borrow gets the capability at the given path, and borrows the capability if it exists.
+        /// Returns `nil` if the capability does not exist or cannot be borrowed using the given type.
+        /// The function is equivalent to `get(path)?.borrow()`.
+        fun borrow<T>(_ path: PublicPath): T?
```

Comment:
`Capability.borrow()` has a constraint on its type parameter: `T: &Any`. Since this is built on top of that mechanism, `Capabilities.borrow` should carry the same type bounds.

Reply:
Great catch!

Both `AuthAccount.borrow` and `Capability.borrow` already have this type bound, so this new `capabilities.borrow` utility function should have it, too.

```suggestion
        fun borrow<T: &Any>(_ path: PublicPath): T?
```